support/ebpf/go_labels.ebpf.c

```diff
@@ -0,0 +1,266 @@
+// This file contains the code for extracting custom labels from Go runtime.
+
+#include "bpfdefs.h"
+#include "kernel.h"
+#include "tracemgmt.h"
+#include "tsd.h"
+#include "types.h"
+#include "util.h"
+
+static inline __attribute__((__always_inline__)) void
+process_value(GoMapBucket *map_value, CustomLabelsArray *out, unsigned i)
+{
+  if (map_value->tophash[i] == 0)
+    return;
+  if (out->len >= MAX_CUSTOM_LABELS)
+    return;
+  CustomLabel *lbl = &out->labels[out->len];
+  if (map_value->keys[i].str != NULL) {
+    unsigned klen = MIN(map_value->keys[i].len, CUSTOM_LABEL_MAX_KEY_LEN - 1);
+    long res      = bpf_probe_read_user(lbl->key, klen, map_value->keys[i].str);
+    if (res) {
+      DEBUG_PRINT(
+        "cl: failed to read key for custom label (%lx): %ld",
+        (unsigned long)map_value->keys[i].str,
+        res);
+      return;
+    }
+    unsigned vlen = MIN(map_value->values[i].len, CUSTOM_LABEL_MAX_VAL_LEN - 1);
+    res           = bpf_probe_read_user(lbl->val, vlen, map_value->values[i].str);
+    if (res) {
+      DEBUG_PRINT("cl: failed to read value for custom label: %ld", res);
+      return;
+    }
+  }
+  out->len++;
+}
+
+static inline __attribute__((__always_inline__)) bool
+process_bucket(PerCPURecord *record, void *label_buckets, int j)
+{
+  CustomLabelsArray *out = &record->trace.custom_labels;
+  GoMapBucket *map_value = &record->goMapBucket;
+  long res =
+    bpf_probe_read_user(map_value, sizeof(GoMapBucket), label_buckets + (j * sizeof(GoMapBucket)));
```

Comment:
Could we read all buckets in one `bpf_probe_read_user` from the parent function or would it be risky to bump into reading past memory page, or needing large amount of scratch buffer space?